server/audit/views.py

```diff
@@ -1,69 +1,202 @@
+            item = audit.inventory_items.get(_id=item_id)
+        except (ObjectDoesNotExist, ValueError) as ex:
+            raise Http404 from ex
+
+        record = {}
+        try: # Check that the item hasn't already been recorded
+            record = Record.objects.get(item_id=item_id)
+        except ObjectDoesNotExist:
+            if item._id in bins.item_ids: # Check that the item belongs to the current bin
+                serializer = self.get_item_serializer(item, many=False)
+                response = Response(serializer.data)
+            else:
+                response = Response({
+                'detail': 'Item part of Audit but not of Bin',
+                'inAudit': audit_id}, status=status.HTTP_400_BAD_REQUEST)
 
-        serializer = self.get_serializer(queryset, many=True)
+        if record and str(record.audit.audit_id) == str(audit_id):
+            response = Response({
+            'detail': 'This item has already been completed',
+            'alreadyMatched': record.record_id},
+            status=status.HTTP_400_BAD_REQUEST)
+
+        return response
+
+    @action(detail=False, methods=['GET'], name='Get Items In Bin')
+    def items(self, request):
+        bin_id = request.query_params.get('bin_id')
+        audit_id = request.query_params.get('audit_id')
+        bins = BinToSK.objects.get(bin_id=bin_id)
+        queryset = Audit.objects.get(audit_id=audit_id)
+        records = Record.objects.filter(bin_to_sk_id=bin_id).all()
+        completed_items =[record.item_id for record in records]
+        bin_items = []
+        for item in queryset.inventory_items.all():
+            if int(item._id) in bins.item_ids and int(item._id) not in completed_items:
+                bin_items.append(item)
+        serializer = self.get_item_serializer(bin_items, many=True)
         return Response(serializer.data)
 
+def set_audit_accuracy(audit_id):
+    audit = Audit.objects.get(audit_id=audit_id)
+    records = Record.objects.filter(audit=audit_id)
+
+    missing = records.filter(status='Missing').count()
+    found = records.filter(status='Provided').count()
+    total_records_no_new = found + missing
 
-class ItemToSKViewSet(viewsets.ModelViewSet):
+    audit_accuracy = 0.0 if total_records_no_new == 0 else found / total_records_no_new
+    setattr(audit, 'accuracy', audit_accuracy)
+    audit.save()
+
+class BinToSKViewSet(viewsets.ModelViewSet):
     """
     API endpoint that allows Audits to be created.
     """
     http_method_names = ['post', 'get', 'delete']
-    queryset = ItemToSK.objects.all()
-    serializer_class = ItemToSKSerializer
+    queryset = BinToSK.objects.all()
+    permission_classes = []
 
     def get_permissions(self):
         factory = PermissionFactory(self.request)
         permission_classes = factory.get_general_permissions([
-            CheckInitAuditData, HasSameOrgInQuery, ValidateSKOfSameOrg])
+            CheckAuditOrganizationById, HasSameOrgInQuery, ValidateSKOfSameOrg])
```

Comment:
Why did you make this change? CheckAuditOrganizationById doesn't even check the right table in the database and now we don't check to make sure the init audit field is part of an audit that is associated with the user's ID.

With these additions, I feel like this class is vulnerable.

Reply:
ill check out what you mean here and get back to you